cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt

```diff
@@ -90,6 +91,66 @@ class StatementHandler(frontend: PythonLanguageFrontend) :
         }
     }
 
+    /**
+     * Translates a Python [`With`](https://docs.python.org/3/library/ast.html#ast.With) into a
+     * [Block].
+     *
+     * We return a Block to handle the with statement, following
+     * [PEP 343](https://peps.python.org/pep-0343/#specification-the-with-statement). The context
+     * managers `__enter__()` runs before the try block and `__exit__()` is always called later in
+     * the `finally` block to clean up, whether an exception occurs or not.
+     */
+    private fun handleWithStatement(node: Python.AST.With): Block {
+        val mainBlock = newBlock().implicit()
+        val exitStatements = mutableListOf<Statement>()
+        val tryBlock = newBlock().implicit()
+
+        node.items.forEach { withItem ->
```

Comment:
In general this is ok for me but we have to agree on this approach because it is a somewhat approximation of the "real" behavior. Basically what you modeled is (with the optional value part):
```python
tmpVar = contextExpr
var = contextExpr.__enter__()
try:
   BODY
finally:
  tmpVar.__exit__()
```

However, in reality the statement behaves like this (adopted from https://peps.python.org/pep-0343/#specification-the-with-statement):
```Python
mgr = contextExpr
tmpVar = mgr.__enter__()
exc = True
try:
    try:
        var = tmpVar
        BLOCK
    except:
        # The exceptional case is handled here
        exc = False
        if not mgr.__exit__(*sys.exc_info()):
            raise
        # The exception is swallowed if exit() returns true
finally:
    # The normal and non-local-goto cases are handled here
    if exc:
        mgr.__exit__(mgr, None, None, None)
```
(the exit calls are actually static calls to the type of the mgr, but we can forego that)

So it is a double-nested try and it seems the intention is to call the exit function with additional information when an exception occurred. Because you probably will not want to call exit again after the exception is handled, the exc check is used and the return code of exit is used to determine if the exception is thrown from the inner try to the outer try (and thus to the block, since the outer try has no catch/except).

Any opinions about how detailed we want to model this? @KuechA @konradweiss @maximiliankaul

Reply:
I'd probably go with a single layer of `try` / `finally`

Reply:
Actually, we can use pretty much what is used in the documentation: https://docs.python.org/3/reference/compound_stmts.html#the-with-statement ;)

Reply:
Done